packages/oas-to-har/src/lib/remove-undefined-objects.js

```diff
@@ -5,26 +5,44 @@ function isEmptyObject(obj) {
 
 // Modified from here: https://stackoverflow.com/a/43781499
 function stripEmptyObjects(obj) {
-  Object.keys(obj).forEach(key => {
-    const value = obj[key];
-    if (typeof value === 'object' && !Array.isArray(obj) && value !== null) {
+  let cleanObj = obj;
+
+  Object.keys(cleanObj).forEach(key => {
+    let value = cleanObj[key];
+
+    if (typeof value === 'object' && !Array.isArray(cleanObj) && value !== null) {
       // Recurse, strip out empty objects from children
-      stripEmptyObjects(value);
+      value = stripEmptyObjects(value);
+
       // Then remove all empty objects from the top level object
       if (isEmptyObject(value)) {
-        // eslint-disable-next-line no-param-reassign
-        delete obj[key];
+        delete cleanObj[key];
+      } else {
+        cleanObj[key] = value;
       }
+    } else if (value === null) {
+      delete cleanObj[key];
     }
   });
+
+  if (Array.isArray(cleanObj)) {
+    // Since deleting a key from an array will retain an undefined value in that array, we need to
+    // filter them out.
+    cleanObj = cleanObj.filter(function (el) {
+      return el !== undefined;
+    });
+  }
+
+  return cleanObj;
 }
 
 function removeUndefinedObjects(obj) {
-  // JSON.stringify removes undefined values
-  const withoutUndefined = JSON.parse(JSON.stringify(obj));
+  // JSON.stringify removes undefined values. Though `[undefined]` will be converted with this to
+  // `[null]`, we'll clean that up next.
+  let withoutUndefined = JSON.parse(JSON.stringify(obj));
 
-  // Then we recursively remove all empty objects
-  stripEmptyObjects(withoutUndefined);
+  // Then we recursively remove all empty objects and nullish arrays.
+  withoutUndefined = stripEmptyObjects(withoutUndefined);
```

Comment:
Is there a reason behind it now using a return value? I think I did it the previous way with mutable values because we weren't copying anything, we were messing with the original variable, so it seems like the truest representation of that.

Reply:
`obj.filter` needs to be assigned to something, and doing a re-assign to `obj` wasn't populating the reference.

<img width="796" alt="Screen Shot 2020-04-28 at 12 49 24 PM" src="https://user-images.githubusercontent.com/33762/80530979-b415bf80-894e-11ea-8861-96ea76ba8418.png">